contracts/FlowCallbackScheduler.cdc

```diff
@@ -0,0 +1,946 @@
+        access(all) let priority: Priority
+        access(all) let executionEffort: UInt64
+        access(all) var status: Status
+
+        /// The timestamp that was requested for this callback
+        /// May be different than the actual scheduled timestamp for medium priority callbacks
+        access(all) let originalTimestamp: UFix64
+
+        /// The actual timestamp that the callback is scheduled for
+        access(all) var scheduledTimestamp: UFix64
+
+        /// Capability to the logic that the callback will execute
+        access(contract) let handler: Capability<auth(ExecuteCallback) &{CallbackHandler}>
+
+        /// Optional data that can be passed to the handler
+        access(contract) let data: AnyStruct?
+
+        /// Fees to pay for the callback
+        access(contract) let fees: @FlowToken.Vault
+
+        access(contract) init(
+            id: UInt64,
+            handler: Capability<auth(ExecuteCallback) &{CallbackHandler}>,
+            data: AnyStruct?,
+            originalTimestamp: UFix64,
+            priority: Priority,
+            executionEffort: UInt64,
+            fees: @FlowToken.Vault,
+            scheduledTimestamp: UFix64
+        ) {
+            self.id = id
+            self.handler = handler
+            self.data = data
+            self.originalTimestamp = originalTimestamp
+            self.priority = priority
+            self.executionEffort = executionEffort
+            self.fees <- fees
+            self.status = Status.Scheduled
+            self.scheduledTimestamp = scheduledTimestamp
+        }
+
+        /// setStatus updates the status of the callback.
+        /// It panics if the callback status is already finalized.
+        access(contract) fun setStatus(newStatus: Status) {
+            pre {
+                self.status != Status.Executed && self.status != Status.Canceled:
+                    "Invalid status: Callback with id \(self.id) is already finalized"
+                newStatus == Status.Executed ? self.status == Status.Processed : true:
+                    "Invalid status: Callback with id \(self.id) cannot be marked as Executed until after it is Processed"
+                newStatus == Status.Processed ? self.status == Status.Scheduled : true:
+                    "Invalid status: Callback with id \(self.id) can only be set as Processed if it is Scheduled"
+            }
+
+            self.status = newStatus
+        }
+
+        /// payAndWithdrawFees withdraws fees from the callback based on the refund multiplier.
+        /// This action is only allowed for canceled callbacks, otherwise it panics.
+        /// It deposits any leftover fees to the FlowFees vault to be used to pay node operator rewards
+        access(contract) fun payAndWithdrawFees(multiplierToWithdraw: UFix64): @FlowToken.Vault {
+            let amount = self.fees.balance * multiplierToWithdraw
+            let feesToReturn <- self.fees.withdraw(amount: amount) as! @FlowToken.Vault
```

Comment:
Since the withdraw amount could be `0.0` when `multiplierToWithdraw == 0.0`, what do you think about calling `createEmptyVault` to avoid the extra `FungibleToken.Withdraw` event emission?

Reply:
That is a good idea. I'll do that